### PR TITLE
Require Python 3.9+ when running bin/tpaexec

### PR DESCRIPTION
**Pull request by amenonsen**
We maintain backwards compatibility with Python 3.6 so far, because both
RHEL7 and RHEL8 have that version by default. But with the cryptography
module removing support for that version, and several pending security
alerts, we want to begin moving away from 3.6 ourselves.

This commit sets the stage for automatically discovering another Python
interpreter (e.g., from edb-python, if available) if the system's
default one is too old. Otherwise, if you initialise your venv with a
different interpreter, or another python is in your PATH, it should just
work sensibly without further ado.

References: TPA-562

### CODE REVIEW

File: bin/tpaexec [hunk 1]
```diff
@@ -1091,10 +1101,33 @@ esac
 # If we can find a venv located relative to TPA_DIR, we activate it as a
 # convenience.
 
+if [ -z ${PYTHON+x} ]; then
```
Comment (amenonsen) [inline bin/tpaexec:1104]:
```suggestion
if [[ -z $PYTHON ]]; then
```

File: bin/tpaexec [hunk 2]
```diff
@@ -1091,10 +1101,33 @@ esac
...
+    PYTHON=/usr/libexec/edb-python39/bin/python3
+    if [ ! -f $PYTHON ]; then
+        PYTHON=python3
+    else
+        echo "Running edb-python $PYTHON"
```
Comment (amenonsen) [inline bin/tpaexec:1109]: Leftover debugging noise?

File: bin/tpaexec [hunk 3]
```diff
@@ -1091,10 +1101,33 @@ esac
 # If we can find a venv located relative to TPA_DIR, we activate it as a
 # convenience.
```
Comment (amenonsen) [inline bin/tpaexec:1102]: Comment has drifted away from the relevant code.

File: bin/tpaexec [hunk 4]
```diff
@@ -1091,10 +1101,33 @@ esac
...
+case "$command" in
+    *help|info|version)
+        # Let through some basic commands, info in particular, without
+        # checking the Python version first.
+        ;;
+
+    *)
+        if [[ $(_python39_or_later) == "False" ]]; then
+            error "TPA requires Python 3.9+, but found only" \
+                "$(_python_version) ($(command -v $PYTHON))"
```
Comment (amenonsen) [inline bin/tpaexec:1126]: Will this break if it couldn't find a python at all?

**Comment by amenonsen** [comment]
NB. Commit needs to be amended to mention 3.9 instead of 3.10

File: bin/tpaexec (hunk 2 above)
Comment (gonzalemario) [reply bin/tpaexec:1109]: No, I thought it was a good idea to have a line like this if we received any bug report.

**Comment by gonzalemario** [comment]
> NB. Commit needs to be amended to mention 3.9 instead of 3.10

I cannot find any reference to 3.10. Can you point it out please?

File: bin/tpaexec (hunk 4 above)
Comment (gonzalemario) [reply bin/tpaexec:1126]: Well. tpa relies on python so it'll be broken if no python is installed. In that case, we'll see an error on the output:

```
$ PYTHON=/bin/non-existing ./bin/tpaexec setup
./bin/tpaexec: line 87: /bin/non-existing: No such file or directory
Creating venv /home/mario/tpa/tpa-venv
./bin/tpaexec: line 119: /bin/non-existing: No such file or directory
ERROR: couldn't initialise /home/mario/tpa/tpa-venv using '/bin/non-existing -m venv'
```

File: bin/tpaexec (hunk 2 above)
Comment (amenonsen) [reply bin/tpaexec:1109]: It's not a good idea to add this output to *every* tpaexec invocation. `tpaexec info` will show the interpreter, that's enough for our purposes.

File: bin/tpaexec (hunk 4 above)
Comment (amenonsen) [reply bin/tpaexec:1126]: I think we could use a better error message there.

Comment (martinmarques) [reply bin/tpaexec:1126]: I don't think that's what Abhijit was referring to. What happens if there is no python installed? What would the $PYTHON variable hold?

File: bin/tpaexec (hunk 2 above)
Comment (gonzalemario) [reply bin/tpaexec:1109]: Done

File: bin/tpaexec (hunk 4 above)
Comment (gonzalemario) [reply bin/tpaexec:1126]: tpaexec will check for edb-python otherwise $PYTHON will be python3 always.

**Comment by JonathanRenon-EDB** [comment]
Should this PR include the doc change to guide user on how to get edb-python39 and remove all the mention to 3.6 as minimum requirement ?

**Comment by amenonsen** [comment]
I've pushed a couple of small fixups. I intend to merge this shortly.

**Review by RichardPooleEDB — APPROVED** [review]
LGTM

**Pull request merged**